HPCC-17381  Improve speed of processHPCCResult function by using map look up

mayx opened the pull request:
- Add a map for case insensitive look up
- Change string compare loop to map look up
- Make all map look ups use the same queryChild function
- No longer add EclName to the map, as EclName is not intended for
  the purpose.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck Hi Tony please review the change.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 3930898617c13ad79ec970573d5f4bfba700dd35
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.408 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |721 |721 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck reviewed (approved):
@mayx looks good

richardkchapman commented on esp/esdllib/esdl_transformer2.cpp:
```diff
@@ -1196,8 +1197,7 @@ void Esdl2Struct::addChildren(Esdl2Transformer *xformer, IEsdlDefObjectIterator
         {
             m_children.append(*obj);
             m_child_map.setValue(obj->queryName(), obj);
-            if (obj->queryEclName())
```
This change (these two lines) is more than just a performance change - it will alter the behaviour.

richardkchapman replied:
Was it intentional, and if so should it be in a separate Jira?

mayx replied:
Yes it's intentional. I checked with Tony that EclName is not supposed to be used this way. We could make it a separate Jira (fix it before applying the optimization).

afishbeck replied:
@mayx Actually thinking about it further, although it wouldn't have been used yet, we should keep EclName support here for future compatibility.

mayx replied:
ok I'll add ecl_name back. Also I found a similar optimization need for EsdlDefStruct, I'll add the fix to the same pull request (unless Tony or Richard think it's better to open another ticket, but I think it's ok to use the same ticket as it's fixing the same problem in 2 areas).

richardkchapman commented:
@mayx, one comment, and a general question - how large is the list of services we iterate through likely to be in typical usage? Are we sure this is not a case of "premature optimization"?

mayx commented:
@richardkchapman the optimization is mostly for recursive/embedded structs (vs. list of services). The list could be big if there's an array of structs. Also the bigger the struct the more the impact.

mayx committed:
HPCC-17381 Improve processHPCCResult speed by using map look up

- Add a map for case insensitive look up
- Change string compare loop to map look up
- Make all map look ups use the same queryChild function

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

mayx commented:
@afishbeck Hi Tony please review the new change, which adds the ecl_name to the maps. I verified that the queryChild function is only used in process functions for response handling, so it's the right approach to add ecl_name to the maps.

mayx commented:
I mentioned similar optimization for EsdlDefStruct. I was thinking to include that in the same ticket, but on second thought, albeit the nature of the change is similar, but that is in a different area, and its queryChild function is only called when processing the request for Java services. So it's probably better to deal with it separately.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: b473fe325f44f0145da005d27018bf17f95639bf
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.07 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |722 |722 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck reviewed (approved):
@mayx looks good.  Did you open a JIRA about making sure ecl_name uniqueness rules are validated at load time / code gen time?  Basically whenever the esdl is used or validated.

mayx commented:
@afishbeck yes a task has been created on that.
